ACPs/224-dynamic-gas-limit-in-subnet-evm/README.md

```diff
@@ -0,0 +1,233 @@
+| :--- | :--- | :--- | :--- |
+| T | target gas consumed per second | 2,000,000 | :white_check_mark: |
+| R | gas capacity added per second | 2*T | :x:
+| C | maximum gas capacity | 10*T | :white_check_mark: Through `MaxCapacityFactor` (default 10)
+| P | minimum target gas consumption per second | 1,000,000 | :x:
+| D | target gas consumption rate update constant | 2^25 | :x:
+| Q | target gas consumption rate update factor change limit | 2^15 | :x:
+| M | minimum gas price | 1 Wei | :white_check_mark:
+| K | initial gas price update factor | ~87*T | :white_check_mark: Through `TimeToDouble` (default 60s)
+
+#### Adjustment to ACP-176 calculations for price discovery
+
+// TODO: MinGasPrice requires a binary search to find the correct value (similar to how we do for DesiredTargetExcess in Coreth)
+
+### Genesis Configuration
+
+There will be a new genesis chain configuration to set the parameters for the chain without requiring the ACP176FeeManager precompile to be activated. This will be similar to the existing fee configuration parameters in chain configuration. If there is no genesis configuration for the new fee parameters the default values for C-Chain will be used. This will look like the following:
+
+```json
+{
+  ...
+  "acp224Timestamp": ...,
+  "acp224FeeConfig": {
+    "targetGas": ...,
+    "minGasPrice": ...,
+    "maxCapacityFactor": ...,
+    "timeToDouble": ...,
+    "targetBlockRate": ...,
+    "minBlockGasCost": ...,
+    "maxBlockGasCost": ...,
+    "blockGasCostStep": ...
+  }
+}
+
+```
+
+### New ACP176 Fee Manager Precompile
+
+A new fee manager precompile will be required to dynamically changing the parameters. The precompile will be similar to the existing fee manager precompile. The solidity interface will be as follows:
+
+```solidity
+//SPDX-License-Identifier: MIT
+pragma solidity ^0.8.24;
+import "./IAllowList.sol";
+
+interface IACP176FeeManager is IAllowList {
+  struct ACP176FeeConfig {
+    uint256 targetGas;
+    uint256 minGasPrice;
+    uint256 maxCapacityFactor;
+    uint256 timeToDouble;
+    uint256 targetBlockRate;
+    uint256 minBlockGasCost;
+    uint256 maxBlockGasCost;
+    uint256 blockGasCostStep;
+  }
+  event FeeConfigChanged(address indexed sender, ACP176FeeConfig oldFeeConfig, ACP176FeeConfig newFeeConfig);
+
+  // Set fee config fields to contract storage
+  function setFeeConfig(
+    uint256 targetGas;
+    uint256 minGasPrice;
+    uint256 maxCapacityFactor;
+    uint256 timeToDouble;
+    uint256 targetBlockRate;
+    uint256 minBlockGasCost;
+    uint256 maxBlockGasCost;
+    uint256 blockGasCostStep;
+  ) external;
+
+  // Get fee config from the contract storage
+  function getFeeConfig()
+    external
+    view
+    returns (
+      uint256 targetGas;
+      uint256 minGasPrice;
+      uint256 maxCapacityFactor;
+      uint256 timeToDouble;
+      uint256 targetBlockRate;
+      uint256 minBlockGasCost;
+      uint256 maxBlockGasCost;
+      uint256 blockGasCostStep;
+    );
```

Comment:
```suggestion
  // Set fee config fields to contract storage
  function setFeeConfig(
    ACP176FeeConfig calldata config
  ) external;

  // Get fee config from the contract storage
  function getFeeConfig()
    external
    view
    returns (
      ACP176FeeConfig memory config
    );
```

Is it possible to just pass the config struct to the precompile interface? If so, I think that's the cleaner approach rather than having 8 individual params

Reply:
I'm not %100 sure but this might end up having to call `encode`/`decode` as follows (in external contract calls)
```
    function updateConfig(FeeConfig memory config) external {
        bytes memory encoded = abi.encode(config);
        feeConfigPrecompile.setFeeConfig(encoded);
    }
```

```
   function readConfig() external view returns (FeeConfig memory config) {
        bytes memory data = feeConfigPrecompile.getFeeConfig();
        config = abi.decode(data, (FeeConfig));
    }
```

Reply:
That shouldn't be necessary, solidity handles the encoding/decoding of parameters for external function calls. 

This will be similar to how the ICM precompile function signatures looks actually: https://github.com/ava-labs/subnet-evm/blob/master/contracts/contracts/interfaces/IWarpMessenger.sol#L36